test/distributed/org/apache/cassandra/distributed/test/sai/PartialUpdateHandlingTest.java

```diff
@@ -0,0 +1,515 @@
+                for (String column : restricted)
+                {
+                    int min = modelRows.get(0).get(column);
+                    clauses.add(column + " >= " + min);
+                    int max = modelRows.get(PARTITIONS_PER_TEST / 2).get(column);
+                    clauses.add(column + " < " + max);
+                }
+            }
+            else
+                throw new IllegalStateException("Validation mode must be EQ or RANGE");
+
+            select.append(String.join(" AND ", clauses));
+
+            InetAddressAndPort node2AddressAndPort = InetAddressAndPort.getByAddress(CLUSTER.get(2).broadcastAddress());
+            Map<String, Index.Status> node2IndexStatusFrom1;
+            Map<String, Index.Status> node2IndexStatusFrom3;
+
+            // Make sure node 2, which is the only complete replica, can't participate in a QUORUM...
+            if (specification.readCL == QUORUM)
+            {
+                node2IndexStatusFrom1 = CLUSTER.get(1).callsOnInstance(() -> IndexStatusManager.instance.peerIndexStatus.remove(node2AddressAndPort)).call();
+                node2IndexStatusFrom3 = CLUSTER.get(3).callsOnInstance(() -> IndexStatusManager.instance.peerIndexStatus.remove(node2AddressAndPort)).call();
+            }
+            else
+            {
+                node2IndexStatusFrom1 = null;
+                node2IndexStatusFrom3 = null;
+            }
+
+            Object[][] result = CLUSTER.coordinator(1).execute(select.toString(), specification.readCL);
+
+            // ...but bring it back up immediately after we make the query:
+            if (specification.readCL == QUORUM)
+            {
+                assertNotNull(node2IndexStatusFrom1);
+                assertNotNull(node2IndexStatusFrom3);
+                CLUSTER.get(1).runOnInstance(() -> IndexStatusManager.instance.peerIndexStatus.put(node2AddressAndPort, node2IndexStatusFrom1));
+                CLUSTER.get(3).runOnInstance(() -> IndexStatusManager.instance.peerIndexStatus.put(node2AddressAndPort, node2IndexStatusFrom3));
+            }
+
+            return result;
+        }
+    }
+
+    @Parameterized.Parameter
+    public Specification specification;
+
+    @Parameterized.Parameters(name = "{0}")
+    public static List<Object[]> parameters()
+    {
+        List<Object[]> parameters = new ArrayList<>();
+
+        // Each test scenario operates over a different set of partition keys. This starting key is the one
+        // used in partition-restricted queries.
+        int partitionKey = 0;
+
+        for (ConsistencyLevel readCL : new ConsistencyLevel[] { ALL, QUORUM })
+        {
+            for (boolean restrictPartitionKey : new boolean[] { false, true })
+            {
+                for (boolean flushPartials : new boolean[] { false, true })
+                {
+                    for (String[] columns : new String[][] { { "ck", "a" }, { "ck", "s" }, { "s", "a" }, { "a", "b" }, { "a" }, { "s" } })
```

Comment:
I missed two static columns here...that might be good to cover, although it doesn't really matter for SAI as much as it does for regular filtering, where that logic is actually different...